Support for custom styles

d4h0 opened the pull request:
Hi,

This implements support for custom styles (required for https://github.com/yaahc/color-eyre/issues/58).

There are a few comments with questions for you, which I have marked with "XXX" (so they are easy to find with grep or ripgrep).

Please let me know if there is any problem or if something could improve. If you want to rename something (English is not my native language), just tell me to do so! :)

The test data are not included at the moment, in case you want to tweak the styles a bit (you probably want, see my "XXX" comments).

After you are ready, you just run the tests in "tests/styles.rs" and follow the instructions to generate the test control data.

d4h0 commented:
@yaahc: So there is a merge conflict.

These PRs, basically, are the first time I used Git for something more complex than "pull/commit/push", so I never had a merge conflict. Do I need to do anything, or will you fix this?

d4h0 commented:
As with the other PR, the test suite will fail because I didn't include the test data.

I will generate and include them now.

Regarding the merge conflict: I looked at the conflict and it looks very simple. But I don't want to resolve it myself (in case I'm missing something and will break something).

yaahc commented:
> @yaahc: So there is a merge conflict.
> 
> These PRs, basically, are the first time I used Git for something more complex than "pull/commit/push", so I never had a merge conflict. Do I need to do anything, or will you fix this?

I can fix the merge conflict, no worries

yaahc commented on src/config.rs:
```diff
@@ -299,6 +473,30 @@ impl HookBuilder {
         }
     }
 
+    // XXX the following tip is a bit hacky, but it's an easy way to test new styles
+
+    /// Set the global styles that `color_eyre` should use.
+    /// 
+    /// Tip: An easy way to test new styles is the following:
+    ///
+    /// 1) clone `color_eyre` via `git clone https://github.com/yaahc/color-eyre.git`
+    /// 
+    /// 2) `cd color-eyre`
+    ///
+    /// 3) open "tests/styles.rs"
+    ///
+    /// 4) search for "# Easy way to test styles"
+    ///
+    /// 5) follow the instructions there
+    ///
+    /// This, basically, leverages the test code to display an error message that includes
+    /// all elements that can be styled. You then can define your own styles which will
+    /// be displayed.
```
I think the recommendation is good but I would tweek it slightly for brevity. I'd want to make it as an example rather than a test and just tell them to check out the test if they want to play with their own style, we already do this a lot with other features that we document in the `README`.

```suggestion
    /// Set the global styles that `color_eyre` should use.
    /// 
    /// **Tip:** You can test new styles by editing `examples/styles.rs` in the `color-eyre` repository.
```

yaahc commented on src/config.rs:
```diff
@@ -612,7 +836,10 @@ struct DefaultPanicMessage;
 impl PanicMessage for DefaultPanicMessage {
     fn display(&self, pi: &std::panic::PanicInfo<'_>, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        writeln!(f, "{}", "The application panicked (crashed).".red())?;
+        // XXX is my assumption correct that this function is guaranteed to only run after `color_eyre` was setup successfully (including setting `STYLES`), and that therefore the following line will never panic? Otherwise, we could return `fmt::Error`, but if the above is true, I like `unwrap` + a comment why this never fails better
```
I'll have to check that this doesn't happen elsewhere but my expectation is that if someone forgets to call the `install` fn then this could cause a double panic. Assuming that's the case I think we'd want to do a `get_or_init` and provide the default style incase it hasn't been setup elsewhere.

d4h0 replied:
My thinking was, that `get_or_init` could lead to confusion here (because suddenly the default theme is used) and that panicking would be better, therefore. `fmt::Error` wouldn't be ideal for the same reason (it's non-descriptive).

> I'll look into that later today to make sure this won't introduce any problems when they're both merged.

Alright. just let me know, what your conclusion was and I'll change this part.

yaahc commented on src/config.rs:
```diff
@@ -6,10 +6,17 @@ use crate::{
+// re-export so end-users don't need to depend on `owo_colors`
+pub use owo_colors::{style, Style};
+
+// XXX Needed to access styles from `impl Section for Report` and `impl PanicMessage for DefaultPanicMessage`
```
:+1:

```suggestion
```

yaahc commented on src/config.rs:
```diff
@@ -22,6 +29,168 @@ impl fmt::Display for InstallError {
+
+impl fmt::Display for InstallStylesError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // XXX I feel this message is not very good – any idea for something better?
```
I think this is perfect personally

```suggestion
```

yaahc commented on src/config.rs:
```diff
@@ -22,6 +29,168 @@ impl fmt::Display for InstallError {
+
+impl fmt::Display for InstallColorBacktraceStylesError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // XXX I feel this message is not very good – any idea for something better?
```
```suggestion
```

yaahc commented on src/config.rs:
```diff
@@ -22,6 +29,168 @@ impl fmt::Display for InstallError {
+
+impl std::error::Error for InstallColorBacktraceStylesError {}
+
+// XXX now that we support text effects, "color scheme" doesn't seem accurate anymore. Let me know if you'd prefer a different name. The good thing about `Styles` is, that it's pretty short. "Theme" would be another good name, I just realized.
```
I think Style is fine, though between `Style` and `Theme` I personally prefer `Theme`.

d4h0 replied:
You are right, I prefer "Theme" as well. I only realized "Theme" would be a good option when I made the last few changes, and I got a bit lazy here... 😅

I will rename "Styles" to "Theme" now.

yaahc commented on src/config.rs:
```diff
@@ -22,6 +29,168 @@ impl fmt::Display for InstallError {
+        }
+    }
+
+    // XXX it would be great, if someone with more style optimizes the light theme below. I just fixed the biggest problems, but ideally there would be darker colors (however, the standard ANSI colors don't seem to have many dark enough colors. Maybe xterm colors or RGB colors would be better (however, again, see my comment regarding xterm colors in `color_spantrace`))
```
I'll try testing it out locally later today, but I'm also not really an expert on these things, all the coloring in this crate is based on the theme from `color-backtrace` :D

d4h0 replied:
👍

I'll keep the "XXX" comments then.

yaahc commented on src/config.rs:
```diff
@@ -34,6 +203,9 @@ pub struct Frame {
     pub lineno: Option<u32>,
     /// source file path
     pub filename: Option<PathBuf>,
+    // XXX is it correct that this is public?
+    /// The global styles that `color_eyre` uses
+    pub styles: Styles,
```
I don't think so. I wouldn't expect us to want to put a `Styles` into the `Frame` type at all. Instead I think we should probably use a `StyledFrame(Frame, Styles)` type that provides the `fmt::Display` implementation we use where we bundle the frame and style together right before printing.

d4h0 replied:
Ah, yes, that makes much more sense (now, that I've implemented it). When I wrote the first version I didn't get why `Frame` is public and assumed users might want to style it themselves for some reason.

yaahc commented on src/config.rs:
```diff
@@ -502,6 +700,19 @@ impl HookBuilder {
     /// Install the given Hook as the global error report hook
     pub fn install(self) -> Result<(), crate::eyre::Report> {
+
+        // XXX I'm not 100% sure if the following is the ideal way to handle these failures (however, I assume it is).
```
As far as I can remember this is how we handle install errors in `eyre` and looks right to me
```suggestion
```

yaahc commented on src/config.rs:
```diff
@@ -502,6 +700,19 @@ impl HookBuilder {
+        // XXX I'm not 100% sure if the following is the ideal way to handle these failures (however, I assume it is).
+
+        if STYLES.set(self.styles).is_err() {
+            // XXX this error can only happen if `install` is called more than one time (because `STYLES` is private)
```
makes sense, at that point I think they'll also run into errors with `eyre`.

yaahc commented:
Okay I went ahead and fixed the merge conflict but It looks like the `owo-colors` stuff doesn't seem to compile, I'm guessing because it still depends on an unreleased version, so I may have introduced some errors. Once the `owo-colors` changes are released to crates make sure to update the `Cargo.toml` to pin the correct minimum version.

d4h0 commented:
Thanks, for the feedback, @yaahc!

In the coming few days I should have enough time to work through your suggestions here and at `color_spantrace`.

d4h0 commented:
Alright, I believe I have implemented all suggestions. Let me know if I missed anything!

I realized I mixed up `color-backtrace` with `color-spantrace` (we had `InstallColorBacktraceStylesError`).

Please keep an eye open for mistakes like these at the public API level... 😅

As with `color_spantrace`, one test fails because I have removed the xterm colors from the default theme (in case you want to compare 'new' and 'old'). I'll generate new test data when a new `owo_colors` version is released.

I have a few additional questions:

1) jam1garner offered to release a new version if we want to publish this earlier (otherwise, it could take some time, until the next version is released).

Should we ask for this?

Personally, I don't care much, but maybe there will be new merge conflicts in the future, otherwise.

If not, and you'd like to properly test my PRs, you could patch your version of `owo_colors` with the current master (my PR is already merged).

2) Is there a better way to add notes, than to add "XXX" comments? I'm not sure if this is the best way to request feedback.

3) `examples/theme_test_helper.rs` could be a bin with optional dependencies. Would that make more sense than an example?

Personally, I think this could confuse people who look at `Cargo.toml` (and it would be more "invasive" in regard to `Cargo.toml`).

yaahc commented:
I'm sorry for not replying to this for so long, life has been getting in the way. Just know that this is high on my list of priorities and I'll get to this as soon as I can.

d4h0 commented:
Don't worry, I was assuming exactly this, and I have enough other stuff to do :)

yaahc reviewed (approved):
I think this looks good, pending review of how the docs look and writing a changelog.

yaahc commented:
ty again for all of this amazing work @d4h0. release `0.5.10` is now out and has your changes in it. :sparkles: :tada: :rocket:, lol

d4h0 commented on tests/theme.rs:
```diff
@@ -1,8 +1,3 @@
-
-
-
-
```
I added these empty lines because in some scenarios the first five lines of a file are extracted for the error (if I remember correctly, I assumed, when color-eyre couldn't find the proper lines, for some reason). In this case, the structure of the error could change when the content of the first five lines changes (at least I was assuming this / I couldn't rule this out completely), which would make the saved test data incompatible.

As far as I remember, this should not be very likely to happen (because we only check ANSI escape sequences), however. Unfortunately, I don't remember the details anymore, so I'm not 100% sure.

I guess rustfmt removed these empty lines. Empty comment lines would be as good as empty lines.

yaahc replied:
I ended up running into too many issues trying to handle testing the output of the source context with `RUST_BACKTRACE=full`, for example, on my machine I'd get the source for `FnOnce`'s frame because I had the `rust-src` component installed but this wasn't on by default in CI so it would not be able to find or display the source code. I think I was able to fix that one but there were other places it still broke that just convinced me to switch to `RUST_BACKTRACE=1` and call it a day. Hopefully that means that now the 5 lines not showing up wont matter?

d4h0 replied:
Ah, I missed that you switched to `RUST_BACKTRACE=1`. Yes, this shouldn't matter much / be okay. The current line is differently styled than the surrounding source lines, but the surrounding lines just were not styled at all (as far as I remember).

d4h0 commented on tests/theme.rs:
```diff
@@ -31,13 +26,10 @@ fn get_error(msg: &'static str) -> Report {
 
-
-
-
```
Same as above. I should have used empty comments and document this better, I've now realized.

d4h0 commented:
Awesome – thank you so much, @yaahc! 😁

I've added two comments, but I guess you got notifications for them.

I hope it wasn't too difficult to finish this. If there is anything I could do better when contributing to a project, please let me know! (I guess, at the very least, I need to learn Git properly... 😅)

yaahc commented:
> I hope it wasn't too difficult to finish this. If there is anything I could do better when contributing to a project, please let me know! (I guess, at the very least, I need to learn Git properly... sweat_smile)

You did fantastic work, and no it wasn't difficult. The test took some thinking to figure out how to get it to work on all our CI configurations but it ended up being simple and working great with all the helper code you setup.

d4h0 commented:
> You did fantastic work, and no it wasn't difficult.

Awesome, thanks! 😄

> The test took some thinking to figure out how to get it to work on all our CI configurations but it ended up being simple and working great with all the helper code you setup.

I think I had similar problems while creating the test (e.g. because of my workspace configuration). Hopefully, this kind of problem will not happen too often... 😅
